hail/src/main/scala/is/hail/io/fs/FS.scala

```diff
@@ -60,7 +70,115 @@ object BGZipCompressionCodec extends CompressionCodec {
   def makeOutputStream(os: OutputStream): OutputStream = new BGzipOutputStream(os)
 }
 
+object FSUtil {
+  def dropTrailingSlash(path: String): String = {
+    if (path.isEmpty)
+      return path
+
+    if (path.last != '/')
+      return path
+
+    var i = path.length - 1
+    while (i > 0 && path(i - 1) == '/')
+      i -= 1
+    path.substring(0, i)
+  }
+}
+
+abstract class FSSeekableInputStream extends InputStream with Seekable {
+  protected[this] var closed: Boolean = false
+  protected[this] var pos: Long = 0
+  protected[this] var eof: Boolean = false
+
+  protected[this] val bb: ByteBuffer = ByteBuffer.allocate(64 * 1024)
+  bb.limit(0)
+
+  def fill(): Unit
+
+  override def read(): Int = {
+    if (eof)
+      return -1
+
+    if (bb.remaining() == 0) {
+      fill()
+      if (eof)
+        return -1
+    }
+
+    pos += 1
+    bb.get().toInt & 0xff
+  }
+
+  override def read(bytes: Array[Byte], off: Int, len: Int): Int = {
+    if (eof)
+      return -1
+
+    if (bb.remaining() == 0) {
+      fill()
+      if (eof)
+        return -1
+    }
+
+    val toTransfer = math.min(len, bb.remaining())
+    bb.get(bytes, off, toTransfer)
+    pos += toTransfer
+    toTransfer
+  }
+
+  def getPosition: Long = pos
+}
+
+abstract class FSPositionedOutputStream extends OutputStream with Positioned {
+  protected[this] var closed: Boolean = false
+  protected[this] val bb: ByteBuffer = ByteBuffer.allocate(64 * 1024)
+  protected[this] var pos: Long = 0
+
+   def flush(): Unit
+
+   def write(i: Int): Unit = {
+    if (bb.remaining() == 0)
+      flush()
+    bb.put(i.toByte)
+    pos += 1
+  }
+
+   override def write(bytes: Array[Byte], off: Int, len: Int): Unit = {
+    var i = off
+    var remaining = len
+    while (remaining > 0) {
+      if (bb.remaining() == 0)
+        flush()
+      val toTransfer = math.min(bb.remaining(), remaining)
+      bb.put(bytes, i, toTransfer)
+      i += toTransfer
+      remaining -= toTransfer
+      pos += toTransfer
+    }
+  }
+
+  def getPosition: Long = pos
+}
+
 trait FS extends Serializable {
+  def containsWildcard(path: String): Boolean = {
```

Comment:
Can you put this on the `FSUtil` object? It's more a helper than part of the API